src/lib.rs

```diff
@@ -0,0 +1,387 @@
+extern crate sha1;
+extern crate sha2;
+extern crate tar;
+extern crate tempfile;
+extern crate toml;
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate serde_json;
+extern crate base_x;
+extern crate handlebars;
+extern crate indexmap;
+extern crate regex;
+extern crate unicode_categories;
+extern crate walkdir;
+extern crate websocket;
+#[macro_use]
+extern crate lazy_static;
+extern crate directories;
+extern crate percent_encoding;
+
+extern crate parity_wasm;
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+extern crate rustc_demangle;
+
+extern crate ansi_term;
+extern crate cargo_metadata;
+
+extern crate memmap;
+extern crate semver;
+
+extern crate atty;
+extern crate open;
+#[macro_use]
+extern crate failure;
+
+mod cargo_shim;
+
+#[macro_use]
+mod utils;
+mod build;
+mod chrome_devtools;
+mod cmd_build;
+mod cmd_deploy;
+mod cmd_prepare_emscripten;
+mod cmd_start;
+mod cmd_test;
+mod config;
+mod deployment;
+mod emscripten;
+mod error;
+mod http_utils;
+mod package;
+mod project_dirs;
+mod test_chromium;
+mod wasm;
+mod wasm_context;
+mod wasm_export_main;
+mod wasm_export_table;
+mod wasm_gc;
+mod wasm_hook_grow;
+mod wasm_inline_js;
+mod wasm_intrinsics;
+mod wasm_js_export;
+mod wasm_js_snippet;
+mod wasm_runtime;
+
+use std::ffi::OsStr;
+use std::net::IpAddr;
+use std::path::PathBuf;
+
+use build::{Backend, BuildArgs};
+pub use cargo_shim::MessageFormat;
+use error::Error;
+use wasm_runtime::RuntimeKind;
+
+/// CLI for `cargo-web`
+#[derive(Debug, StructOpt)]
+#[structopt(name = "cargo-web")]
+#[structopt(about = "A `cargo` subcommand for the client-side web.")]
+#[structopt(raw(global_setting = "structopt::clap::AppSettings::ColoredHelp"))]
+#[structopt(raw(setting = "structopt::clap::AppSettings::VersionlessSubcommands"))]
+#[structopt(rename_all = "kebab-case")]
+pub enum CargoWebOpts {
+    /// Compile a local package and all of its dependencies
+    Build {
+        #[structopt(flatten)]
+        build_args: Build,
+        #[structopt(flatten)]
+        ext: BuildExt,
+        #[structopt(flatten)]
+        build_target: Target,
+    },
+    /// Typecheck a local package and all of its dependencies
+    Check {
+        #[structopt(flatten)]
+        build_args: Build,
+        #[structopt(flatten)]
+        ext: BuildExt,
+        #[structopt(flatten)]
+        build_target: Target,
+    },
+    /// Deploys your project so that it's ready to be served statically
+    Deploy {
+        /// Output directory; the default is `$CARGO_TARGET_DIR/deploy`
+        #[structopt(short = "o", long, parse(from_os_str))]
+        output: Option<PathBuf>,
+        #[structopt(flatten)]
+        build_args: Build,
+    },
+    /// Fetches and installs prebuilt Emscripten packages
+    PrepareEmscripten,
+    /// Runs an embedded web server, which serves the built project
+    Start {
+        /// Bind the server to this address
+        #[structopt(long, parse(try_from_str), default_value = "127.0.0.1")]
```

Comment:
Any reason you've changed it to `127.0.0.1` from the previous default value of `localhost`?

Reply:
`localhost` doesn't successfully parse as a `std::net::IpAddr` - it does work with `ToSocketAddrs` (the original impl) because [the name resolution is done implicitly there](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html).

I can add a trivial-ish fn that does the name resolution if you prefer using the hostname.

Reply:
> `localhost` doesn't successfully parse as a `std::net::IpAddr` - it does work with `ToSocketAddrs` (the original impl) because [the name resolution is done implicitly there](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html).
> 
> I can add a trivial-ish fn that does the name resolution if you prefer using the hostname.

Oh, okay, so now only IP addresses are supported. Yes, please add name resolution back in case someone depends on it (I doubt it, but I do take backwards compatibility seriously when I can).